# Raw Read Quality Assessment

It's extremely important to do quality assessment and improvement
steps before working with sequencing data. I highly recommend this one
take an iteractive approach to this, consisting of diagnostics with
`qrqc`, then improvement, then more diagnostics. This iterative
approach has two benefits:

1. It prevents the dangerous assumption that tools that work well
generally are working well with a particular dataset. 

2. Extremely pathological datasets stick out more in comparative
before/after quality improvement software diagnostics.

```{r setup}
opts_chunk$set(fig.width=7, fig.height=7, cache=TRUE)
opts_knit$set(base.url="https://github.com/vsbuffalo/rna-seq-example/raw/master/")
````

## Load Required Packages and Set Number of Cores to Use

```{r load-packages}
library(reshape)
library(ggplot2)
library(multicore)
library(qrqc)

options(mc.cores=4)
````

## Raw Quality Reports

```{r summarize-raw-reads}

raw.fastq.files <- list.files("data/raw-reads", pattern=".*\\.fastq", full.names=TRUE)
names(raw.fastq.files) <- basename(raw.fastq.files)
raw.fastq.summaries <- mclapply(raw.fastq.files, readSeqFile)
````

### Base Quality

```{r raw-base-quality}

qualPlot(raw.fastq.summaries)

````

### Base Frequency

```{r raw-base-frequency}

basePlot(raw.fastq.summaries)

````

### K-mer Contaminant Plots

```{r raw-kmer-kl}

kmerKLPlot(raw.fastq.summaries)

````

### Entropy Contaminant Plots

```{r raw-entropy}

kmerEntropyPlot(raw.fastq.summaries)

````

## Statistic of Processed Reads (after Sickle and Scythe)

These statistics were gathered from `scythe` and `sickle` output. This
provides a look how many reads were removed, the distribution of the
adapter contaminants.


```{r read-in-trimming-data}

d <- read.table("data/scythe-data.txt", sep="\t", header=TRUE)

````

This is a bit of a hodgepodge of a data file; it has original read
counts, counts at each stage, and a character column of Scythe output
indicating where the contaminants were found. First, we extract this
last column. I plan on changing Scythe soon (and perhaps Sickle) so
that output is nicer for downstream statistics.

```{r remove-contaminant-col}
d.scythe <- local({
  # look at just adapter columns, and the id (file) col
  tmp <- d[, c(1, grep("adapter", colnames(d))), ]
  
  # split out and convert the values to numeric
  tmp.a1 <- lapply(strsplit(as.character(tmp[, 2]), ", "), as.numeric)
  tmp.a2 <- lapply(strsplit(as.character(tmp[, 3]), ", "), as.numeric)
  
  # append a position vector
  tmp.a1[[length(tmp.a1)+1]] <- seq_along(tmp.a1[[1]])
  tmp.a2[[length(tmp.a2)+1]] <- seq_along(tmp.a2[[1]])
  
  # create a long dataframe of the above data for each adapter, 
  # and rbind both together
  a1 <- data.frame(adapter="adapter 1", do.call(cbind, tmp.a1))
  colnames(a1)[2:6] <- c(as.character(tmp[, 1]), "position")
  a2 <- data.frame(adapter="adapter 2", do.call(cbind, tmp.a2))
  colnames(a2)[2:6] <- c(as.character(tmp[, 1]), "position")
  rbind(a1, a2)
})

d.scythe <- melt(d.scythe, id.vars=c("adapter", "position"))

p <- ggplot(d.scythe) + geom_bar(aes(x=position, y=value, fill=adapter), position="dodge", stat="identity")
p <- p + scale_y_continuous("count")
p
````

Now we can look at how many reads Scythe found to be contaminated:

```{r}

d.scythe.trimmed <- melt(d[, c("file", "total", "uncontaminated")], id.vars=c("file"))

p <- ggplot(d.scythe.trimmed) + geom_bar(aes(x=file, y=value, fill=variable)) 
p <- p + scale_y_continuous("count")
p
````

And how many reads Sickle trimmed:

```{r}

d.sickle <- melt(d[, c("file", "kept", "discarded")], id.vars=c("file"))

p <- ggplot(d.sickle) + geom_bar(aes(x=file, y=value, fill=variable)) 
p <- p + scale_y_continuous("count")
p
````

## Post-Processed Quality Reports

```{r summarize-processed-reads}

processed.fastq.files <- list.files("data/improved-reads", pattern=".*\\.fastq", full.names=TRUE)
names(processed.fastq.files) <- basename(processed.fastq.files)
processed.fastq.summaries <- mclapply(processed.fastq.files, readSeqFile)
````

### Base Quality

```{r processed-base-quality}

qualPlot(processed.fastq.summaries)

````

### Base Frequency

```{r processed-base-frequency}

basePlot(processed.fastq.summaries)

````

### K-mer Contaminant Plots

```{r processed-kmer-kl}

kmerKLPlot(processed.fastq.summaries)

````

### Entropy Contaminant Plots

```{r processed-entropy}

kmerEntropyPlot(processed.fastq.summaries)

````

